diesel/src/pg/connection/mod.rs

```diff
@@ -1,113 +1,223 @@
+use bytes::BytesMut;
+use postgresql::fallible_iterator::FallibleIterator;
+use postgresql::types::Type;
+use postgresql::Statement;
+use std::cell::RefCell;
+use std::convert::TryFrom;
+use std::error::Error;
+
+/// This connection type is based on the pure rust postgresql implementation
+/// inside the [postgres](https://docs.rs/postgres/0.17.0/postgres/index.html) crate.
+///
+/// See the documentation on [`postgres::Client`](https://docs.rs/postgres/0.17.0/postgres/struct.Client.html#method.connect)
+/// for connection strings that are accepted by `PgConnection::establish`
+/// or use the the `TryFrom` impl to create a new `PgConnection` from
+/// an existing `postgres::Client`
 #[allow(missing_debug_implementations)]
 pub struct PgConnection {
-    pub(crate) raw_connection: RawConnection,
+    conn: RefCell<postgresql::Client>,
     pub(crate) transaction_manager: AnsiTransactionManager,
     statement_cache: StatementCache<Pg, Statement>,
     metadata_cache: PgMetadataCache,
 }
 
-unsafe impl Send for PgConnection {}
+impl PgConnection {
+    /// Get a reference to the inner `postgres::Client`
+    pub fn inner_connection(&self) -> &RefCell<postgresql::Client> {
+        &self.conn
+    }
+}
+
+impl TryFrom<postgresql::Client> for PgConnection {
+    type Error = ConnectionError;
+
+    fn try_from(client: postgresql::Client) -> Result<PgConnection, ConnectionError> {
+        let conn = Self {
+            conn: RefCell::new(client),
+            transaction_manager: AnsiTransactionManager::new(),
+            statement_cache: StatementCache::new(),
+            metadata_cache: PgMetadataCache::new(),
+        };
+
+        conn.set_config_options()
+            .map_err(CouldntSetupConfiguration)?;
+        Ok(conn)
+    }
+}
 
 impl SimpleConnection for PgConnection {
     fn batch_execute(&self, query: &str) -> QueryResult<()> {
-        let query = CString::new(query)?;
-        let inner_result = unsafe { self.raw_connection.exec(query.as_ptr()) };
-        PgResult::new(inner_result?)?;
+        self.conn.borrow_mut().simple_query(query)?;
         Ok(())
     }
 }
 
 impl Connection for PgConnection {
-    type Backend = Pg;
+    type Backend = super::Pg;
     type TransactionManager = AnsiTransactionManager;
 
-    fn establish(database_url: &str) -> ConnectionResult<PgConnection> {
-        RawConnection::establish(database_url).and_then(|raw_conn| {
-            let conn = PgConnection {
-                raw_connection: raw_conn,
-                transaction_manager: AnsiTransactionManager::new(),
-                statement_cache: StatementCache::new(),
-                metadata_cache: PgMetadataCache::new(),
-            };
-            conn.set_config_options()
-                .map_err(CouldntSetupConfiguration)?;
-            Ok(conn)
-        })
+    fn establish(database_url: &str) -> ConnectionResult<Self> {
+        use postgresql::tls::NoTls;
+
+        let client = postgresql::Client::connect(database_url, NoTls)
```

Comment:
How do we handle tls vs no tls, a feature flag ?

Reply:
For now I would assume `NoTLS` for the generic `establish` method. For using the `Tls` support from the postgres crate (or any other connection specific stuff from there) there is a `TryFrom<postgres::Client>` impl somewhere, so that you can construct a customized connection on your own.